engine/src/test/java/org/camunda/bpm/engine/test/bpmn/event/timer/BoundaryTimerNonInterruptingEventTest.java

```diff
@@ -755,6 +762,54 @@ public void shouldExecuteTimeoutListenerJobOnOrAfterDueDate() {
     assertThat(timerJob.getDuedate()).isEqualTo(timerDueDate);
   }
 
+  @Test
+  @Deployment
+  public void shouldReevaluateTimerCycleWhenDue() throws Exception {
+    // given
+    MyBean myBean = new MyBean("R2/PT1H");
+    processEngineConfiguration.getBeans().put("myBean", myBean);
+    processEngineConfiguration.setReevaluateTimeCycleWhenDue(true);
+
+    runtimeService.startProcessInstanceByKey("nonInterruptingCycle").getId();
+
+    JobQuery jobQuery = managementService.createJobQuery();
+    assertThat(jobQuery.count()).isEqualTo(1);
+    moveByHours(1); // execute first job
+    assertThat(jobQuery.count()).isEqualTo(1);
+
+    // when bean changed and job is due
+    myBean.setCycle("R2/PT2H");
+    moveByHours(1); // execute second job
+
+    // then one more job is left due in 2 hours
+    assertThat(jobQuery.count()).isEqualTo(1);
+    assertThat(jobQuery.singleResult().getDuedate())
+        .isEqualToIgnoringMinutes(new Date(ClockUtil.getCurrentTime().getTime() + ((2 * 60 * 1000 * 60))));
+  }
+
+  @Test
+  @Deployment(resources = {"org/camunda/bpm/engine/test/bpmn/event/timer/BoundaryTimerNonInterruptingEventTest.shouldReevaluateTimerCycleWhenDue.bpmn20.xml"})
+  public void shouldNotReevaluateTimerCycleWhenDue() throws Exception {
+    // given
+    MyBean myBean = new MyBean("R2/PT1H");
+    processEngineConfiguration.getBeans().put("myBean", myBean);
+    processEngineConfiguration.setReevaluateTimeCycleWhenDue(false);
+
+    runtimeService.startProcessInstanceByKey("nonInterruptingCycle").getId();
+
+    JobQuery jobQuery = managementService.createJobQuery();
+    assertThat(jobQuery.count()).isEqualTo(1);
+    moveByHours(1); // execute first job
+    assertThat(jobQuery.count()).isEqualTo(1);
+
+    // when bean changed and job is due
+    myBean.setCycle("R2/PT2H");
+    moveByHours(1); // execute second job
+
+    // then no more jobs are left (two jobs has been executed already)
+    assertThat(managementService.createJobQuery().singleResult()).isNull();
+  }
+
   //we cannot use waitForExecutor... method since there will always be one job left
   private void moveByHours(int hours) throws Exception {
```

Comment:
🔧 How about we rewrite this a bit using the `testRule` to simply execute existing jobs instead of starting the job executor and letting the main thread sleep?

Reply:
As the comment above mentions, there will always be one job left. As more test cases use repeatable timers in this test class. `testRule#executeAvailableJobs()` also can't be used due to the recursion inside of it.

Reply:
`#executeAvailableJobs(false)` disables the recursion, executes only the currently available job, so it works for this test class. Will adjust the helper method.